File: android/src/main/java/com/lightningkite/kwift/views/SelectDateRangeMonthCVD.shared.kt
package com.lightningkite.kwift.views

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.util.DisplayMetrics
import android.view.View
import com.lightningkite.kwift.observables.*
import com.lightningkite.kwift.rx.addWeak
import com.lightningkite.kwift.time.*
import java.util.*

/**Renders a swipeable calendar.**/
open class SelectDateRangeMonthCVD : MonthCVD() {
    override fun generateAccessibilityView(): View? = null

    var draggingStart: Boolean = true
    var start: MutableObservableProperty<DateAlone?> = StandardObservableProperty(null)
    var endInclusive: MutableObservableProperty<DateAlone?> = StandardObservableProperty(null)
    init {
        this.start.onChange.addWeak(this) { self, value ->
            self.invalidate()
        }
        this.endInclusive.onChange.addWeak(this) { self, value ->
            self.invalidate()
        }
    }

    val selectedDayPaint: Paint = Paint()
    val selectedPaint: Paint = Paint()

    override fun measure(width: Float, height: Float, displayMetrics: DisplayMetrics) {
        super.measure(width = width, height = height, displayMetrics = displayMetrics)
        selectedDayPaint.textSize = dayPaint.textSize
    }

    val drawDay_dateAlone: DateAlone = DateAlone(0,0,0)
    override fun drawDay(
        canvas: Canvas,
        showingMonth: DateAlone,
        day: DateAlone,
        displayMetrics: DisplayMetrics,
        outer: RectF,
        inner: RectF
    ) {
        when {
            day == start.value && (day == endInclusive.value || endInclusive.value == null) -> {
                CalendarDrawing.dayBackground(canvas, inner, selectedPaint)
                CalendarDrawing.day(canvas, showingMonth, day, inner, selectedDayPaint)
            }
            day == start.value -> {
                CalendarDrawing.dayBackgroundStart(canvas, inner, outer, selectedPaint)
                CalendarDrawing.day(canvas, showingMonth, day, inner, selectedDayPaint)
            }
            day == endInclusive.value -> {
                CalendarDrawing.dayBackgroundEnd(canvas, inner, outer, selectedPaint)
                CalendarDrawing.day(canvas, showingMonth, day, inner, selectedDayPaint)
            }
            day.comparable > (start.value?.comparable ?: Int.MAX_VALUE) && day.comparable < (endInclusive.value?.comparable ?: Int.MIN_VALUE) -> {
                CalendarDrawing.dayBackgroundMid(canvas, inner, outer, selectedPaint)
                CalendarDrawing.day(canvas, showingMonth, day, inner, selectedDayPaint)
            }
            else -> {
                CalendarDrawing.day(canvas, showingMonth, day, inner, dayPaint)
            }
        }
    }


    private var startedDraggingOn: DateAlone? = null

    override fun onTap(day: DateAlone) {
        if(start.value != null && start.value == endInclusive.value && day.comparable > start.value!!.comparable){
            endInclusive.value = day
        } else {
            start.value = day
            endInclusive.value = day
        }
    }

    override fun onTouchDown(day: DateAlone): Boolean {
        if(day != start.value && day != endInclusive.value){
            return false
        }
        startedDraggingOn = day
        //If on start/end - drag
        //If after, extend
        //If before, extend
        //If middle, collapse all
        val startValue = start.value
        val endInclusiveValue = endInclusive.value

        when {
            startValue == null || endInclusiveValue == null -> {
                start.value = day
                endInclusive.value = day
                draggingStart = false
            }
            day == endInclusiveValue -> {
                draggingStart = false
            }
            day == startValue -> {
                draggingStart = true
            }
            day.comparable > endInclusiveValue!!.comparable && startValue == endInclusiveValue -> {
                endInclusive.value = day
                draggingStart = false
            }
            else -> {
                start.value = day
                endInclusive.value = day
                draggingStart = false
            }
        }
        return true
    }

    override fun onTouchMove(day: DateAlone): Boolean {
        val startValue = start.value
        val endInclusiveValue = endInclusive.value
        when {
            startValue == null || endInclusiveValue == null -> {
            }
            draggingStart && day.comparable > endInclusiveValue!!.comparable -> {
                start.value = endInclusive.value
                endInclusive.value = day
                draggingStart = false
                return true
            }
            !draggingStart && day.comparable < startValue!!.comparable -> {
                endInclusive.value = start.value
                start.value = day
                draggingStart = true
                return true
            }
        }

        val obs: MutableObservableProperty<DateAlone?> = if (draggingStart) start else endInclusive
        obs.value = day
        return true
    }

    override fun onTouchUp(day: DateAlone): Boolean {
        onTouchMove(day)
        return true
    }
}

File: ios/Khrysalis/com/lightningkite/khrysalis/views/SelectDateRangeMonthCVD.shared.swift
//Package: com.lightningkite.khrysalis.views
//Converted using Khrysalis2

import Foundation
import RxSwift
import RxRelay



open class SelectDateRangeMonthCVD: MonthCVD {
    
    
    
    override public func generateAccessibilityView() -> View?  {
        return nil
    }
    public var draggingStart: Bool
    public var start: MutableObservableProperty<DateAlone?>
    public var endInclusive: MutableObservableProperty<DateAlone?>
    public var selectedDayPaint: Paint
    public var selectedPaint: Paint
    
    override public func measure(width: Float, height: Float, displayMetrics: DisplayMetrics) -> Void {
        super.measure(width: width, height: height, displayMetrics: displayMetrics)
        selectedDayPaint.textSize = dayPaint.textSize
    }
    override public func measure(_ width: Float, _ height: Float, _ displayMetrics: DisplayMetrics) -> Void {
        return measure(width: width, height: height, displayMetrics: displayMetrics)
    }
    public var drawDay_dateAlone: DateAlone
    
    override public func drawDay(canvas: Canvas, showingMonth: DateAlone, day: DateAlone, displayMetrics: DisplayMetrics, outer: RectF, inner: RectF) -> Void {
        if day == start.value && ( day == endInclusive.value || endInclusive.value == nil ) {
            CalendarDrawing.dayBackground(canvas, inner, selectedPaint)
            CalendarDrawing.day(canvas, showingMonth, day, inner, selectedDayPaint)
        } else if day == start.value {
            CalendarDrawing.dayBackgroundStart(canvas, inner, outer, selectedPaint)
            CalendarDrawing.day(canvas, showingMonth, day, inner, selectedDayPaint)
        } else if day == endInclusive.value {
            CalendarDrawing.dayBackgroundEnd(canvas, inner, outer, selectedPaint)
            CalendarDrawing.day(canvas, showingMonth, day, inner, selectedDayPaint)
        } else if day.comparable > ( start.value?.comparable ?? Int32.MAX_VALUE ) && day.comparable < ( endInclusive.value?.comparable ?? Int32.MIN_VALUE ) {
            CalendarDrawing.dayBackgroundMid(canvas, inner, outer, selectedPaint)
            CalendarDrawing.day(canvas, showingMonth, day, inner, selectedDayPaint)
        } else  {
            CalendarDrawing.day(canvas, showingMonth, day, inner, dayPaint)
        }
    }
    override public func drawDay(_ canvas: Canvas, _ showingMonth: DateAlone, _ day: DateAlone, _ displayMetrics: DisplayMetrics, _ outer: RectF, _ inner: RectF) -> Void {
        return drawDay(canvas: canvas, showingMonth: showingMonth, day: day, displayMetrics: displayMetrics, outer: outer, inner: inner)
    }
    private var startedDraggingOn: DateAlone? 
    
    override public func onTap(day: DateAlone) -> Void {
        if start.value != nil, start.value == endInclusive.value, day.comparable > start.value!.comparable {
            endInclusive.value = day
        } else {
            start.value = day
            endInclusive.value = day
        }
    }
    override public func onTap(_ day: DateAlone) -> Void {
        return onTap(day: day)
    }
    
    override public func onTouchDown(day: DateAlone) -> Bool {
        if day != start.value, day != endInclusive.value {
            return false
        }
        startedDraggingOn = day
        var startValue = start.value
        var endInclusiveValue = endInclusive.value
        if startValue == nil || endInclusiveValue == nil {
            start.value = day
            endInclusive.value = day
            draggingStart = false
        } else if day == endInclusiveValue {
            draggingStart = false
        } else if day == startValue {
            draggingStart = true
        } else if day.comparable > endInclusiveValue!.comparable && startValue == endInclusiveValue {
            endInclusive.value = day
            draggingStart = false
        } else  {
            start.value = day
            endInclusive.value = day
            draggingStart = false
        }
        return true
    }
    override public func onTouchDown(_ day: DateAlone) -> Bool {
        return onTouchDown(day: day)
    }
    
    override public func onTouchMove(day: DateAlone) -> Bool {
        var startValue = start.value
        var endInclusiveValue = endInclusive.value
        if startValue == nil || endInclusiveValue == nil {
            } else if draggingStart && day.comparable > endInclusiveValue!.comparable {
            start.value = endInclusive.value
            endInclusive.value = day
            draggingStart = false
            return true
        } else if !draggingStart && day.comparable < startValue!.comparable {
            endInclusive.value = start.value
            start.value = day
            draggingStart = true
            return true
        }
        var obs: MutableObservableProperty<DateAlone?> = {if draggingStart {
            return start
        } else {
            return endInclusive
        }}()
        obs.value = day
        return true
    }
    override public func onTouchMove(_ day: DateAlone) -> Bool {
        return onTouchMove(day: day)
    }
    
    override public func onTouchUp(day: DateAlone) -> Bool {
        onTouchMove(day)
        return true
    }
    override public func onTouchUp(_ day: DateAlone) -> Bool {
        return onTouchUp(day: day)
    }
    
    override public init() {
        let draggingStart: Bool = true
        self.draggingStart = draggingStart
        let start: MutableObservableProperty<DateAlone?> = StandardObservableProperty(nil)
        self.start = start
        let endInclusive: MutableObservableProperty<DateAlone?> = StandardObservableProperty(nil)
        self.endInclusive = endInclusive
        let selectedDayPaint: Paint = Paint()
        self.selectedDayPaint = selectedDayPaint
        let selectedPaint: Paint = Paint()
        self.selectedPaint = selectedPaint
        let drawDay_dateAlone: DateAlone = DateAlone(0, 0, 0)
        self.drawDay_dateAlone = drawDay_dateAlone
        let startedDraggingOn: DateAlone?  = nil
        self.startedDraggingOn = startedDraggingOn
        super.init()
        self.start.onChange.addWeak(self) { (self, value) in 
            self.invalidate()
        }
        self.endInclusive.onChange.addWeak(self) { (self, value) in 
            self.invalidate()
        }
    }
}
 
